### PR TITLE
Train UNet-3D

### PR DESCRIPTION
**Pull request by gml16**
- Implementation of UNet-3D architecture and training. The architecture follows [the one implemented in MLPerf](https://github.com/mlcommons/training/blob/master/image_segmentation/pytorch/model/unet3d.py).
  - Includes implementation of Dice and Cross Entropy loss.
  - Adds support for `output_size` in `ConvTranspose2d`, this allows to dynamically pad upsampled convolution. It also follows the [PyTorch API](https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html). The `num_spatial_dims` argument also makes it dimension-agnostic (i.e. it becomes `ConvTransposeNd`). I'd be happy to generalise `Conv2d` and `ConvTranspose2d` into `ConvNd` and `ConvTransposeNd` in a future PR if that is of interest.
- The data is automatically downloaded when running training with `python -m examples.train_unet_3d`. 

There are 2 big caveats with this PR:
- I am not able to run a full training locally as my old GPU does not have enough memory (and it crashes on Collab due to exceeded RAM). However, I did train with the same images cropped and the Dice score is increasing steadily. I would be very grateful if someone could try training and report the results.
- I cannot find the labels for the test dataset on the [KIT dataset webpage](https://kits19.grand-challenge.org/data/). If anyone knows where to find it and can point me to it that would be very much appreciated.

It's my first PR in this repo, feedback is more than welcome :)

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/nn/__init__.py                   18               2
tinygrad/tensor.py                         1               1
------------------------------------------------------------
total                                     19               3
------------------------------------------------------------
lines added in the tinygrad folder: 16
```

**Comment by terafo** [comment]
How much VRAM would your implementation require to run unclipped training? Why existing UNet-3D code and kits19 dataloader aren't used?

**Comment by gml16** [comment]
Thank you @terafo for taking a look. 

The VRAM requirement shouldn't be very high. The batch size is 1 as the images have variable sizes, the largest one being 1059x512x512 voxels. This is 1110MB in float32 without gradients. I am not entirely sure the gradient overhead introduced by tinygrad, if we triple it that would be below 4GB (which is the VRAM of my laptop GPU, so it may be more).

Regarding your comment about reusing UNet-3D and kits19 I am not sure if you mean open-source code outside of tinygrad? I haven't seen such implementation within tinygrad, but I may have missed it. Otherwise, I did follow the KITS data loader [here](https://github.com/neheller/kits19/blob/master/starter_code/get_imaging_v2.py). And I followed  the architecture [implemented in MLPerf](https://github.com/mlcommons/training/blob/master/image_segmentation/pytorch/model/unet3d.py) but it is written in PyTorch, not in tinygrad.

**Comment by chenyuxyz** [comment]
https://github.com/search?q=repo%3Atinygrad%2Ftinygrad%20unet3d&type=code

**Comment by gml16** [comment]
Thank you for pointing out the very obvious 🤦 I somehow had only included tinygrad folder in my search
